torch/nested/_internal/ops.py

```diff
@@ -1260,6 +1260,34 @@ def _nested_get_max_seqlen(func, *args, **kwargs):
     inp = new_kwargs.pop("input")
     return inp._metadata_cache.get("max_seqlen", None)
 
+@register_jagged_func(torch.ops.aten.masked_select.default, "self: jt, mask: jt")
+def masked_select_default(func, *args, **kwargs):
+    _, new_kwargs = normalize_function(
+        func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
+    )
+    inp = new_kwargs.pop("input")
+    mask = new_kwargs.pop("mask")
+
+    mask_values = None
+    if mask._offsets.equal(inp._offsets):
+        mask_values = mask.values()
+    elif len(mask._values) == len(inp._offsets) - 1:
+        mask_values = torch.repeat_interleave(
+            mask.values(),
+            (inp._lengths if inp._lengths is not None else torch.diff(inp._offsets)),
+        )
+    else:
+        raise RuntimeError(
+            f"Jagged mask with offsets {mask.offsets()} is not compatible with input's offsets {inp.offsets()}"
+        )
```

Comment:
Some nits on shape checking:
- Always check that the ndim of both args is 2 (I see that this is the only case tested, I think its okay if we only support this case for now if its all we need)
- If we want to check that the raggedness is the same, we only need to compare the shapes. If you print out the shape of a nested tensor [B, j0], the special int j0 represents the particular pattern of raggedness for that NJT. Checking the data is fine for cpu in eager, but in compile this leads to data dependent control flow. 
- For the two shapes to broadcast its not enough to check just that the len are equal. We also need to check that `[B, j0] x [B, 1]` the second dimension of the mask must be 1. This also means that the mask must NOT be a nested tensor [B, j1] because we would not be able to inspect the actual ragged pattern of j1 without incurring a graph break. 

```python
if inp.ndim != 2 or mask.ndim != 2:
    raise RuntimeError(...)
if inp.shape == mask.shape:
    mask_values = mask._values
# [B, j0] x [B, 1]
elif inp.shape[0] == mask.shape[0] and mask.shape[1] == 1:
    mask_values = ...
else:
    raise RuntimeError(...)
```

Reply:
> For the two shapes to broadcast its not enough to check just that the len are equal. We also need to check that [B, j0] x [B, 1] the second dimension of the mask must be 1. This also means that the mask must NOT be a nested tensor [B, j1] because we would not be able to inspect the actual ragged pattern of j1 without incurring a graph break.

So for 2D nested tensors (which we're going to limit to for this initial implementation), there are two main scenarios I can think of for applying the select with a JT / T combination as a first iteration:

Given a nested tensor input (NT) and a mask m (Tensor) 
1) input.values() has the number of elements as m
The mask would be applied equivalently to 

input.values().masked_select(m)

2) The number of segments in the JT tensor ( len(input.lengths() ) is equal to the number of elements in the mask 

In which care we apply the boolean per segment in the JT

So an example:
input.values() == [1,2,3,4,5,6,7,8]
input.offsets() == [0,3,5,8]
input.lengths() == [3,2,3]

Applying a mask of [True, False, True] would zero out the elements of the middle segment.

out.values() == [1,2,3,6,7,8]
out.offsets() == [0,3,3,8]
out.lengths() == [3,0,3]

This is similar to what TF offers. https://www.tensorflow.org/api_docs/python/tf/ragged/boolean_mask#examples

I think the second use case is not something we immediately need so we can hold off on implementing this.